### DISCUSSION TITLE
TypeScript support! 🚀

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#457] [Announcements]
beta.62 brings with it a long-awaited, much-requested feature: TypeScript support!

![CleanShot 2021-09-27 at 10 42 38](https://user-images.githubusercontent.com/36073/134951810-31754840-85c3-4ad3-a493-59c757fdda07.png)

### 1. But, how?

Each time your run a TS script, Script Kit will compile the TS script using `esbuild` to a JS script in a `.scripts` dir (notice the "dot"). The compiled JS script is then imported from there. Using `.scripts` as a sibling dir will help avoid any `import`/path issues. You can also write TS "library" files in your `~/.kenv/lib` dir and import them into your script just fine.

If you're experienced with `esbuild` and curious about the settings, they look like this:

```js
let { build } = await import("esbuild")

await build({
  entryPoints: [scriptPath],
  outfile,
  bundle: true,
  platform: "node",
  format: "esm",
  external: ["@johnlindquist/kit"],
})
```

This also opens the door to exporting/building/bundling scripts and libs as individual shippable tools which I'll investigate more in the future.

### 2. Can I still run my JS scripts if I switch to TS?

Yes! Both your TS and JS scripts will show up in the UI.

### 3. Why the `import "@johnlindquist/kit"`?

When you create a new TS script, the generated script will start with the line: `import "@johnlindquist/kit"`

This is mostly to make your editor stop complaining by forcing it to load the type definition files and forcing it to treat the file as an "es module" so support "top-level `await`". It's not technically required since it's not technically importing anything, but your editor will certainly complain very loudly if you leave it out.

### 4. Where is the setting stored?

Look in your `~/.kenv/.env` for `KIT_MODE=ts`.

## fs-extra's added to global

The [fs-extra methods](https://www.npmjs.com/package/fs-extra#methods) are now added on the global space. I found myself using `outputFile`, `write/readJson`, etc too often and found them to be a great addition. The only one missing is `copy` since we're already using that to "copy to clipboard". You can bring it in with the normal import/alias process if needed, e.g., `let {copy:fsCopy} = await import("fs-extra")`

## Sync Path

![CleanShot 2021-09-27 at 11 10 26](https://user-images.githubusercontent.com/36073/134954703-7c9d779f-268a-4f8b-973a-59ac71eebaf0.png)

You may notice running scripts from the Script Kit app that some commands you can run in your terminal might be missing, like "yarn", etc.

Run the following command in your terminal to copy the $PATH var from your terminal to your `~/.kenv/.env`. This will help "sync" up which commands are available between your terminal and running scripts from the app.

```bash
~/.kit/bin/kit sync-path
```

**Comment by kentcdodds** [comment]
What's my tsconfig need to look like for this to work? I think this was added for me:

```json
{
  "compilerOptions": {
    "module": "esnext",
    "target": "esnext",
    "moduleResolution": "node"
  }
}
```

But typescript can't seem to find anything 🤷‍♂️

![image](https://user-images.githubusercontent.com/1500684/135193945-b177053e-7b3d-4c10-b43e-985fa97faf63.png)

**Comment by johnlindquist** [comment]
I'm pushing out an update in ~1 hour or so that should clean that up. The previous build was supposed to update the `~/.kenv/package.json` with a dep of 

```json
"@johnlindquist/kit": "file:../.kit"
```

but the patch script had a tiny bug that skipped that step.

It was one of those things that worked fine on a clean install, but had this hiccup when patching/updating.

**Reply by johnlindquist** [reply]
Please update and let me know if it's sorted 🙏

**Reply by kentcdodds** [reply]
That sorted it!

One interesting thing is TypeScript didn't like this:

```ts
const dir = await env(
  'SHORTEN_REPO_DIRECTORY',
  'Where is your netlify-shortener repo directory?',
)
```

And I had to change it to this:

```ts
const dir = await env('SHORTEN_REPO_DIRECTORY', () =>
  Promise.resolve('Where is your netlify-shortener repo directory?'),
)
```

That seemed odd...

Also this:

![image](https://user-images.githubusercontent.com/1500684/135204951-eaef9fde-082f-4add-8f25-68d1b73027fe.png)

**Reply by johnlindquist** [reply]
Great, thanks. I'll dive into those tomorrow and report back.

Going to add a bunch of tsdoc "@example"s to so they show up in the hints

**Reply by kentcdodds** [reply]
Stellar
